src/hotspot/share/c1/c1_ValueMap.hpp

```diff
@@ -187,7 +187,13 @@ class ValueNumberingVisitor: public InstructionVisitor {
   void do_Convert        (Convert*         x) { /* nothing to do */ }
   void do_NullCheck      (NullCheck*       x) { /* nothing to do */ }
   void do_TypeCast       (TypeCast*        x) { /* nothing to do */ }
-  void do_NewInstance    (NewInstance*     x) { /* nothing to do */ }
+  void do_NewInstance    (NewInstance*     x) {
+    ciInstanceKlass* c = x->klass();
+    if (c != nullptr && !c->is_initialized() &&
+        ((c->is_loaded() && c->has_class_initializer()) || !c->is_loaded())) {
```

Comment:
```suggestion
        (!c->is_loaded() || c->has_class_initializer()) {
```

Reply:
Thank you for this simplification. I forgot about short-circuiting in this case...